src/mscorlib/src/System/Array.cs

```diff
@@ -1559,31 +1559,23 @@ public static void Reverse<T>(T[] array)
         {
             if (array == null)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
-            Reverse(array, 0, array.Length);
+
+            // The Span ctor might perform an unnecessary type check if T is a reference type,
+            // so we use MemoryMarshal.CreateSpan to suppress this check.
+
+            MemoryMarshal.CreateSpan(ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData()), array.Length).Reverse();
         }
 
         public static void Reverse<T>(T[] array, int index, int length)
         {
             if (array == null)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
-            if (index < 0)
-                ThrowHelper.ThrowIndexArgumentOutOfRange_NeedNonNegNumException();
-            if (length < 0)
-                ThrowHelper.ThrowLengthArgumentOutOfRange_ArgumentOutOfRange_NeedNonNegNum();
-            if (array.Length - index < length)
+            if ((uint)index > (uint)array.Length)
+                ThrowHelper.ThrowArgumentOutOfRange_IndexException();
```

Comment:
This will change exception messages. The same argument validation is in non-generic Array.Reverse - we should keep the two in sync.

Reply:
Is this necessary? MSDN [explicitly documents](https://msdn.microsoft.com/en-us/library/z78xtwts(v=vs.110).aspx) the specific exception types thrown by the non-generic method, and I'd rather not make changes to the parameter validation for these older methods. At least with the generic method the particular exception types aren't explicitly documented, so we have a little more leeway to go between `ArgumentException` and `ArgumentOutOfRangeException` in order to simplify the validation logic.

Reply:
I have not realized that this is changing exception types too. Have you gone through the exercise on how breaking is that?

Reply:
This code previously threw `ArgumentOutOfRangeException` if _length_ was negative. Now it throws `ArgumentException`.

This _shouldn't_ break anybody, and in the past when I've participated in full framework shiproom we wouldn't have considered this a breaking change. (The reason it wouldn't be considered a breaking change is that we're not introducing a code path that now throws where previously it didn't, and we're still throwing an exception type that would have been called out in the documentation were docs to exist.)

If you have specific guidelines I should go against send them along and I'll take a look.

Reply:
The breaking change rules are documented in https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#exceptions . Your change is throwing less specific exception in some cases. This is not listed as allowed change.

Also, note you typically pick up the new generic overloads silently with recompilation. It makes the behavior divergence between the non-generic and generic overloads more questionable.

Reply:
I agree that the probability of breaking anybody here is low. However, we are not doing paper cut breaks unless the benefit is significant.

Reply:
Ok, reverted.